### DISCUSSION TITLE
Location-Based usable yet ?

### DISCUSSION THREAD

**Discussion by CarloCattano** [AR-js-org/AR.js#412] [Help]
Hello there . Im having a lot of trouble getting purely location based experiences to work .
This library seems awesome and would love to create a project around it . However , when trying any of the given examples in the repo , AR js Studio etc . The content seems either to stick to the camera (as mentioned in some issues ) , or scaling not working at all : Content looks the same size no matter the distance to the location (three js ) 

I would love to have an easy way to implement a kind of tour experience where info (images + audio ) spawn according to the location of the users using mobile phones, but as it is now , this doesnt seem doable.

Im reading the issues , tutorials , and testing all posible examples and the results are always similar . 

Is there something Im overlooking here ? 

Im aware that gps precision is not the best but I can do with around 10m error without problem , but as I said as of now , its very time consuming to develop and Im not getting anywhere.

I kindly appreciate any help on the matter .
I've tried also the suggested gps coords page to double check the coords , console logged them with alerts , updated on every gps update etc and still no luck

**Comment by CarloCattano** [comment]
this is what Im trying today . Is there a way to make the text scale according to distance ?
My intent is to trigger callbacks to show Images , text and an audio if the user gets close to the designated place , I have to try maxDistance etc but would love to know if there is a good way to do this before I dive into it .


```html 
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title>GeoAR.js demo</title>
    <script src="https://aframe.io/releases/1.0.4/aframe.min.js"></script>
    <script src="https://unpkg.com/aframe-look-at-component@0.8.0/dist/aframe-look-at-component.min.js"></script>
    <script src="https://raw.githack.com/AR-js-org/AR.js/master/aframe/build/aframe-ar-nft.js"></script>
  </head>

  <body>
    <a-scene
      vr-mode-ui="enabled: false"
      arjs="sourceType: webcam; videoTexture: true; debugUIEnabled: false;"
    >
      <a-text
        value="HOA"
        look-at="[gps-camera]"
        scale="100 100 100"
        gps-entity-place="latitude: 00.49792262177317; longitude: 00.429618712294344;"
      ></a-text>
      <a-text
        value="Hospital"
        look-at="[gps-camera]"
        scale="100 100 100"
        gps-entity-place="latitude: 00.496978793988816;longitude: 00.428899880305176;"
      ></a-text>
      <a-text
        value="Cafe"
        look-at="[gps-camera]"
        scale="100 100 100" , 
        gps-entity-place="latitude: 00.497680951543465; longitude: 00.428921337981912;"
      ></a-text>
      <a-image
        src="assets/1.jpeg"
        look-at="[gps-camera]"
        scale="2 2 2" ,
        gps-entity-place="latitude: 00.497528479682366; longitude: 00.427194167474749;"
  ></a-image>
      <a-camera gps-camera rotation-reader> </a-camera>
    </a-scene>
  </body>
</html> 
```

**Reply by nickw1** [reply]
As I said in the other issue, try out the three.js location-based example from latest `dev`. This should show four boxes a short distance from you: red to the north, yellow to the south, blue to the west, and green to the east.

Note it will only work on Android/Chrome, due to the lack of iOS devices for testing.

We are proposing to replace the A-Frame components with new ones (see #406) which have simpler code and thus should be easier to debug.

**Reply by CarloCattano** [reply]
Thanks a lot @nickw1, I will do that and report back.
This is such and exciting project 
I will try to get a working example by cloning the PR or the dev branch as you say

**Comment by kalwalt** [comment]
@CarloCattano you are welcome! We would improve this part as possible if you can test new features, will boost the development.

**Reply by CarloCattano** [reply]
Will do that ASAP and ask for feedback on Apple devices around

**Comment by CarloCattano** [comment]
dev branch seems to be up to date with master "This branch is up to date with master."
I will try with the enhancement pull request you suggested instead

**Reply by CarloCattano** [reply]
the new-location based example yields blank screen on Android mobile (Chrome latest) , seems to open the camera on PC but when testing on android I get a blank screen (white) , GPS is logged properly . 
See my test at  [<https://github.com/CarloCattano/webXRtest/tree/main/newtest>](<https://github.com/CarloCattano/webXRtest/tree/main/newtest>)

Just removed the location simulation , and only one warning is present on PC 
```
Error with Permissions-Policy header: Unrecognized feature: 'interest-cohort'.
```

Must find a good way to debug mobile console somehow

**Reply by nickw1** [reply]
Did you try the pure three.js example? (no A-Frame)

https://github.com/AR-js-org/AR.js/tree/master/three.js/examples/location-based

PR #406 (which needs tweaking to reflect recent changes in dev) wraps the three.js location based API so would be good to see if the pure three version is working for you first.

You can use Chrome Developer Tools to debug on mobile, e.g. see https://developer.chrome.com/docs/devtools/remote-debugging/

**Reply by CarloCattano** [reply]
Im trying it now , it seems to work . Even tho I cant understand how to make the scaling more "realistic". 
I assume the gps position update is being evaluated but need to make a proper test outdoors.
[https://github.com/CarloCattano/webXRtest/tree/main/three](https://github.com/CarloCattano/webXRtest/tree/main/three)

**Reply by nickw1** [reply]
Great! If the scaling is not realistic it's probably to do with the camera parameters (field of view, etc) not exactly matching those of your device. Try replacing the vertical field of view in the example with the actual value for your device, if you know it. Note it's a vertical, not horizontal field of view, you can get the vertical field of view by dividing the horizontal fov by the aspect ratio.

**Comment by nickw1** [comment]
@CarloCattano I've fixed some bugs with the new A-Frame location based example (#406) so feel free to try that again.

There is also a live example on the web at
https://hikar.org/narjs/AR.js/aframe/examples/new-location-based/

**Reply by CarloCattano** [reply]
this is what I get 
![image](https://user-images.githubusercontent.com/17380530/160237357-591a1e68-21a1-4a83-b491-bfe2aa30d784.png)

**Reply by CarloCattano** [reply]
it might be that Im importing things wrong , I copied the raw modified builds manually after your pushes tho 
https://github.com/CarloCattano/webXRtest/tree/main/newtest

**Reply by nickw1** [reply]
Does my live example at hikar.org work https://hikar.org/narjs/AR.js/aframe/examples/new-location-based/?

It works for me on both desktop and mobile. The bug you're getting I corrected (the MathUtils bug) in the latest commits for #406.

If you're still getting errors, try clearing the cache/storage for Chrome on your mobile device. It could be loading an old version from the cache, with the error still there.

**Reply by CarloCattano** [reply]
Thanks! For some reason it works on brave (my usual browser) but chrome fails, even after deleting the cache, will try to force quit it

**Reply by CarloCattano** [reply]
Ok , its working now as intended . Will take it outdoors for a test . Seems smoother than the three js approach even.
Made some adjustments to work as I imagine it would approach a tour experience 
```javascript
window.onload = () => {
    const el = document.querySelector("[gps-new-camera]");
    el.addEventListener("gps-camera-update-position", e => {
      
            alert(`Got first GPS position: lon ${e.detail.position.longitude} lat ${e.detail.position.latitude}`);
            const properties = [{
                    color: 'red',
                    latDis: 52.49792262177317,
                    lonDis: 13.429618712294344
                },{
                    color: 'yellow',
                    latDis: 52.496978793988816,
                    lonDis: 13.42889988030517
                },{
                    color: 'blue',
                    latDis: 52.497680951543465,
                    lonDis: 13.428921337981912
                },{
                    color: 'green',
                    latDis: 52.497528479682366,
                    lonDis: 13.427194167474749
                }
            ];
            for(const prop of properties) {
                const entity = document.createElement("a-box");
                entity.setAttribute("scale", {
                    x: 20, 
                    y: 20,
                    z: 20
                });
                entity.setAttribute('material', { color: prop.color } );
                entity.setAttribute('gps-new-entity-place', {
                    latitude:  prop.latDis,
                    longitude:  prop.lonDis
                });
                document.querySelector("a-scene").appendChild(entity);
            }
    });
};
```

**Reply by nickw1** [reply]
@CarloCattano glad it's working!

**Reply by cramvisuals** [reply]
Hey! Not sure if this is the place for this but I'm an apple user here testing stuff out. Read through all of pull #406 and am so excited to see this issue getting attention! 

I just visited the live hikar example and input my local coordinates on these devices:
- 2021 MacBook Pro, MacOS Monterey 12.2.1, Brave browser v1.37.109
- iPhone 12 pro max, iOS 15.3.1, Chrome app v100.0.4896.85 (most recent version as of this comment)

The computer camera activates and red box shows up just fine on camera, but doesn't move.
On iPhone, the camera opens and input box retrieves my location, but nothing shows up. I'll take a walk outside with it and see if that changes anything.

**Reply by CarloCattano** [reply]
Brave can be a bit unpredictable, with all the security included. Try to deactivate de brave shield too

**Reply by phatpixels** [reply]
Hi All,
I came here looking for examples & clues. I have been  exploring AR.js for a number of weeks now. I cant seem to get an AR location based example to work. I'm trying to get some simple text to appear as you approach a local monument in my town. I would like the GPS location of the monument to trigger the text appearing on your phone when in close proximity to the monument. I have used the basic AR.js example swapped out the GPS coordinates & text with my own. The text does appear, but it always appears no matter where you are, even if miles away. I think i might be missing something here, but im not sure what. I've tested on various devices, ive tested these examples posted here. The examples function fine, but, they are constant. I wish them to only work at the specific GPS location. I will continue with exploring webAR with passion. Looking forward to any small tips along the way. Many thanks.

**Reply by nickw1** [reply]
@phatpixels which specific version are you using? Are you using A-Frame or three.js? Are you using the `new-location-based` or `location-based` components?

**Reply by phatpixels** [reply]
@nickw1 Hi, Firstly i tried the basics of the location based example here at https://ar-js-org.github.io/AR.js-Docs/ it worked fine on the spot spawning based on current camera location,  Ignoring the specific GPS location i had inserted.

I then found this https://github.com/AR-js-org/AR.js/tree/master/aframe/examples/new-location-based it worked fine out of the box spawning from current camera location, then i tried adding the monument GPS location details and nothing appeared. 

The basic first basic example here worked similarly on both android and ipad although ignoring the specified gps details appearing only from current camera location.

I love where webar is going, the library is so great! So many wonderful use cases for location AR. I was hoping to start promoting the town with webAR, starting with a simple monument example that works equally well across both apple and android devices. I'll continue testing, there seems to be various configurations on git. Reading the comments here i'm a little confused as to which direction to go for my needs. But i will persist.

**Reply by nickw1** [reply]
@phatpixels ok, that's the latest version so should work fine.

Do you have your code? Wondering if it's anything to do with your monument model and/or the text being too small. (Apologies if this sounds too obvious a problem but it's best to check this first).

**Reply by phatpixels** [reply]
@nickw1 thanks i agree it pays to check all aspects. I need to spend more time testing this weekend, im using exactly this code unchanged except for the GPS location, testing with the 4 coloured boxes (n,s,e,w)  https://github.com/AR-js-org/AR.js/tree/master/aframe/examples/new-location-based 
Its nice and stable, i just need it to trigger only when i reach the GPS spot. Im positive the issue is my end, Testing it on apple devices also, do you know if it runs on apple currently?

**Reply by nickw1** [reply]
Ah, it could be an issue with Apple devices, the new API is known not to work properly on Apple at the moment. You could try the old `location-based` API and see if you have more luck with that. One issue is that none of the developers currently have an Apple device which means that testing is limited to Android devices or desktop setups. I do know that Apple has more quirks than Android.

Sorry about that, though I think one of the developers (@kalwalt) is trying to obtain an Apple device which will obviously help. This is why we would welcome developers who own Apple devices and who can help with testing!

**Comment by garthhenning** [comment]
I have a rough prototype of our Running Reality historical map augmented reality system up using the newly revised AR.js location-based system. It definitely took some work to get it converted over, but it seems to be mostly working. I've been testing on an iOS device and can get the orientation sensor to activate mostly reliably. You can activate the augmented reality view by pushing the "LIVE" button anywhere on our map if you are on a mobile device and at zoom level 16 or deeper, such as here: [RR First Battle of Manassas](https://www.runningreality.org/#07/21/1861&38.81608,-77.52134&zoom=16)
